libs/agno/agno/db/sqlite/sqlite.py

```diff
@@ -0,0 +1,1832 @@
+
+        elif table_type == "evals":
+            if not hasattr(self, "eval_table"):
+                if self.eval_table_name is None:
+                    raise ValueError("Eval table was not provided on initialization")
+
+                self.eval_table = self._get_or_create_table(table_name=self.eval_table_name, table_type="evals")
+
+            return self.eval_table
+
+        else:
+            raise ValueError(f"Unknown table type: '{table_type}'")
+
+    def _get_or_create_table(self, table_name: str, table_type: str) -> Table:
+        """
+        Check if the table exists and is valid, else create it.
+
+        Args:
+            table_name (str): Name of the table to get or create
+            table_type (str): Type of table (used to get schema definition)
+
+        Returns:
+            Table: SQLAlchemy Table object
+        """
+        with self.Session() as sess, sess.begin():
+            table_is_available = is_table_available(session=sess, table_name=table_name)
+
+        if not table_is_available:
+            return self._create_table(table_name=table_name, table_type=table_type)
+
+        # SQLite version of table validation (no schema)
+        if not is_valid_table(db_engine=self.db_engine, table_name=table_name, table_type=table_type):
+            raise ValueError(f"Table {table_name} has an invalid schema")
+
+        try:
+            table = Table(table_name, self.metadata, autoload_with=self.db_engine)
+            log_debug(f"Loaded existing table {table_name}")
+            return table
+
+        except Exception as e:
+            log_error(f"Error loading existing table {table_name}: {e}")
+            raise
+
+    # -- Session methods --
+
+    def _upsert_agent_session_raw(self, session: AgentSession) -> Optional[dict]:
+        """
+        Insert or update an agent session in the database.
+
+        Args:
+            session (AgentSession): The session data to upsert.
+
+        Returns:
+            Optional[dict]: The upserted session, or None if operation failed.
+
+        Raises:
+            Exception: If an error occurs during upsert.
+        """
+        try:
+            table = self._get_table(table_type="sessions")
+
+            # Calculated fields
+            chat_history = (
+                json.dumps([chat_message.to_dict() for chat_message in session.chat_history])
+                if session.chat_history
+                else None
+            )
+            runs = json.dumps([run.to_dict() for run in session.runs]) if session.runs else None
+            summary = json.dumps(session.summary.to_dict()) if session.summary else None
+
+            # Convert other JSON fields to strings
+            agent_data = json.dumps(session.agent_data) if session.agent_data else None
+            session_data = json.dumps(session.session_data) if session.session_data else None
+            extra_data = json.dumps(session.extra_data) if session.extra_data else None
```

Comment:
Like this... feels like this should be done inside the session so you can have less "business logic" here.

Reply:
Here we cast dicts to json strings because of sqlite. I will encapsulate it in a serialization function instead of in the model itself